cpp/src/arrow/scalar.cc

```diff
@@ -194,4 +200,138 @@ Status CheckBufferLength(const FixedSizeBinaryType* t, const std::shared_ptr<Buf
 }
 }  // namespace internal
 
+// CastImpl(...) assumes `to` points to a non null scalar of the correct type with
+// uninitialized value
+
+// error fallback
+Status CastImpl(const Scalar& from, Scalar* to) {
+  return Status::NotImplemented("casting scalars of type ", *from.type, " to_type type ",
+                                *to->type);
+}
+
+// numeric to numeric
+template <typename From, typename To>
+Status CastImpl(const NumericScalar<From>& from, NumericScalar<To>* to) {
+  to->value = static_cast<typename To::c_type>(from.value);
+  return Status::OK();
+}
+
+// numeric to boolean
+template <typename T>
+Status CastImpl(const NumericScalar<T>& from, BooleanScalar* to) {
+  constexpr auto zero = static_cast<typename T::c_type>(0);
+  to->value = from.value != zero;
+  return Status::OK();
+}
+
+// boolean to numeric
+template <typename T>
+Status CastImpl(const BooleanScalar& from, NumericScalar<T>* to) {
+  to->value = static_cast<typename T::c_type>(from.value);
+  return Status::OK();
+}
+
+// string to any
+template <typename ScalarType>
+Status CastImpl(const StringScalar& from, ScalarType* to) {
+  std::shared_ptr<Scalar> out;
+  RETURN_NOT_OK(Scalar::Parse(to->type, util::string_view(*from.value), &out));
+  to->value = std::move(checked_cast<ScalarType&>(*out).value);
+  return Status::OK();
+}
+
+// binary to string
+template <typename T>
+Status CastImpl(const BinaryScalar& from, StringScalar* to) {
+  to->value = from.value;
+  return Status::OK();
+}
+
+// numeric to string
+template <typename T>
+Status CastImpl(const NumericScalar<T>& from, StringScalar* to) {
+  to->value = Buffer::FromString(std::to_string(from.value));
+  return Status::OK();
+}
+
+// boolean to string
+Status CastImpl(const BooleanScalar& from, StringScalar* to) {
+  to->value = Buffer::FromString(from.value ? "true" : "false");
+  return Status::OK();
+}
+
+template <typename ToType>
+struct UnpackFromType {
```

Comment:
Why is it called "unpack"?

Reply:
would you prefer "unbox"?

Reply:
Hmm... well, in both cases I don't understand what's being unpacked or unbox :-)

Reply:
The destination or original type is being unboxed from `DataType` to `NullType, BooleanType, ...`. I thought the terms 'unpack' and 'unbox' were fairly universal for that; what would you call "down cast object based on a discriminant"? I could just call them both Impl, if you prefer.

Reply:
Hmm... Perhaps `FromCastTypeVisitor` and `ToTypeCastVisitor`?